src/definitions/x509/crl.rs

```diff
@@ -0,0 +1,319 @@
+        Some(distribution_points) => distribution_points,
+    };
+
+    let mut cert_lists = vec![];
+
+    for distribution_point in distribution_points.0.iter() {
+        let distribution_point_name = distribution_point.distribution_point.as_ref().ok_or(
+            Error::DistributionPointMalformed("missing distributionPoint name"),
+        )?;
+
+        let urls = distribution_point_urls(distribution_point_name)?;
+
+        for url in urls.iter() {
+            let crl_bytes = fetch_crl(url).await?;
+
+            cert_lists.push(validate_crl(root_cert, &crl_bytes)?);
+        }
+    }
+
+    Ok(cert_lists)
+}
+
+fn distribution_point_urls(name: &DistributionPointName) -> Result<Vec<String>, Error> {
+    match name {
+        DistributionPointName::FullName(uris) => {
+            let uris: Result<Vec<String>, Error> = uris
+                .iter()
+                .map(|general_name| match general_name {
+                    GeneralName::UniformResourceIdentifier(s) => Ok(s.to_string()),
+                    _ => Err(Error::DistributionPointMalformed(
+                        "distribution point name something other than URI",
+                    )),
+                })
+                .collect();
+
+            uris
+        }
+        DistributionPointName::NameRelativeToCRLIssuer(_) => Err(
+            Error::DistributionPointMalformed("contained relative to issuer name"),
+        ),
+    }
+}
+
+async fn fetch_crl(url: &str) -> Result<Vec<u8>, Error> {
+    let bytes = reqwest::get(url)
+        .await
+        .map_err(Error::FetchingCrl)?
+        .bytes()
+        .await?;
+
+    Ok(bytes.to_vec())
+}
+
+fn read_distribution_points(cert: &TbsCertificate) -> Result<Option<CrlDistributionPoints>, Error> {
+    let extensions = match cert.extensions.as_ref() {
+        None => return Ok(None),
+        Some(extensions) => extensions,
+    };
+
+    let crl_extension = match extensions
+        .iter()
+        .find(|ext| ext.extn_id == CrlDistributionPoints::OID)
+    {
+        None => return Ok(None),
+        Some(crl_extension) => crl_extension,
+    };
+
+    let mut der_reader = SliceReader::new(crl_extension.extn_value.as_bytes())
+        .map_err(|err| Error::ParsingCrlComponent("extension", err))?;
+
+    let distribution_points = CrlDistributionPoints::decode(&mut der_reader)
+        .map_err(|err| Error::ParsingCrlComponent("extension", err))?;
+
+    Ok(Some(distribution_points))
+}
+
+fn validate_crl(cert: &TbsCertificate, crl_bytes: &[u8]) -> Result<TbsCertList, Error> {
+    let (crl_raw, crl) = decode_cert_list(crl_bytes)?;
+
+    if cert.issuer != crl.tbs_cert_list.issuer {
+        return Err(Error::IssuerMismatchBetweenCertAndCrl);
+    }
+
+    if cert.subject_public_key_info.algorithm != crl.signature_algorithm {
+        return Err(Error::SignatureTypeMismatch(
+            Box::new(cert.subject_public_key_info.algorithm.clone()),
+            Box::new(crl.signature_algorithm),
+        ));
+    }
```

Comment:
I'm going to respond to this and the comment below simultaneously.  That comment copied below for posterity:

> @tristanmiller-spruceid could you please address the outstanding issues?

> The CRL is verified against the certificate subject's public key, instead of the certificate issuer. This only works if the certificate being checked is self-signed (e.g. the root). The current API does not support checking the CRL of an intermediate certificate, unless it happens that the intermediate certificate and the root certificate have the CRL.

> To resolve this please consider implementing the public API as described in the issue:
> https://github.com/spruceid/isomdl/issues/76#issue-2014541461

First, given the complexity of certificate chains, I've split the public API of the crl crate into two halves, 1) given any cert in the chain, grab the CRLs that this cert signs and check that CRL's validity (fetch_and_validate_crl) and 2) given a CRL and a cert, return if the cert is revoked (check_cert_against_cert_lists).  This allows for relatively arbitrary certificate chains as implemented by higher level code where there may be several layers of certificate chains.

So it's expected that calling code will iterate through their cert tree chain, attach all CRLs to that chain each grabbed with fetch_and_validate_crl, and validate each node in that chain against each of the child CRLs with check_cert_against_cert_lists.

Additionally, the CRLs are signed by the subject public key of the the cert who's distribution point specified them.  That does not stop from checking the CRL of an intermediate certificate, one simply needs to call fetch_and_validate_crl against the intermediate certs as well.  The argument to fetch_and_validate_crl has had it's name changed from root_cert to crl_signing_cert to make that more clear.

Reply:
> Additionally, the CRLs are signed by the subject public key of the the cert who's distribution point specified them. 

Correct me if I'm wrong, but I believe CRLs are signed by the issuer of the cert whose distribution point specified them, as you pointed out:

> RFC 5280 6.3.1. CRL Processing (b)(1)
>>    [V]erify that the CRL issuer matches the certificate issuer.

For a root CA, this happens to be itself as it is self-signed.

> That does not stop from checking the CRL of an intermediate certificate, one simply needs to call fetch_and_validate_crl against the intermediate certs as well. The argument to fetch_and_validate_crl has had it's name changed from root_cert to crl_signing_cert to make that more clear.

The problem is that for an intermediate certificate the `fetch_and_validate_crl` function will fail because the intermediate certificate does not sign its own CRL, the CRL issuer is the certificate issuer.